### PR TITLE
rfc: bnorm: add support for bwd_data and bwd_wei

### PR DESCRIPTION
**Pull request by bartekxk**
# Description

This RFC describes a proposal of support backward_data (unsupported with scale and shift) and backward_weights in batch normalization.

[Rendered document.](https://github.com/oneapi-src/oneDNN/blob/bkocot/rfc/bnorm_data/rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md)

### CODE REVIEW

File: rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
+# Proposal for support backward data (actually not supported for scale and shift) and backward weights in batch normalization.
+
+## Problem and motivation
+
+Currently in OneDNN library batch normalization support following propagation 
+kinds: forward_inference, forward_training, backward and backward_data (only 
+without scale and shift). Tensorflow team requested the possibility to use 
+backward_data with scale and shift. It is worth adding backward_weights
```
Comment (dzarukin) [inline rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md:8]: > Tensorflow team requested the possibility to use 
backward_data with scale and shift.

Do we know what the problem is to use `dnnl_backward` in such case? It does what is requested.

**Review by dzarukin — CHANGES_REQUESTED** [review]

File: rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
+# Proposal for support backward data (actually not supported for scale and shift) and backward weights in batch normalization.
+
+## Problem and motivation
+
+Currently in OneDNN library batch normalization support following propagation 
+kinds: forward_inference, forward_training, backward and backward_data (only 
+without scale and shift). Tensorflow team requested the possibility to use 
+backward_data with scale and shift. It is worth adding backward_weights 
+propagation kind in this case as well. It will create full support for 
+possible all propagation kinds in batch normalization.
```
Comment (kwiersch) [inline rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md:10]: nit: `possible all propagation kinds` -> `all possible propagation kinds`

File: rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+| D ss             | calculate data                                        |
+| D                | calculate data and weights, dont save scale and shift |
+| W                | Not supported                                         |
+
+The implementation may just split the current one into two sections (calculate 
+weights and calculate data) and execute it depending on propagation kind.
+
+### Option 2
+Option two is the same as the first except no support for backward_weights. 
+Only the requested part will be implemented. This will reduce the
```
Comment (kwiersch) [inline rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md:60]: In this option, how would the requested parts be determined? Is it by what memory objects are passed at runtime?

File: rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md (hunk 2 above)
Comment (bartekxk) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md]: Done.

File: rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md (hunk 3 above)
Comment (bartekxk) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md]: I meant that requested by user - backward_weights. Changed for better readability.

Comment (kwiersch) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md]: It's still not clear to me. _How_ does user indicate what parts are requested? Is it by which memory objects are passed at execution time?

Comment (bartekxk) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md]: It can be passed by descriptor. Like for example here: https://github.com/intel-innersource/libraries.performance.math.onednn/blob/aeebe74d6b87e10be658865f25eacd9486994ae9/examples/primitives/batch_normalization.cpp#L107 User can pass `backward_data` or `backward_weights`.

Comment (kwiersch) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md]: Okay, I really do not understand this option. I thought that "only the requested part will be implemented" meant that user could somehow indicate which part of outputs is desired (ie `diffsrc` or `diffgamma` and `diffbeta`), but without relying on the propagation kind. Is that right? Or am I misinterpreting what is meant here by "user" -- do you mean the specific user that requested this feature, or a generic user of the API?

Either way, it would help to be more precise about what Option 2 entails. Thanks!

Comment (bartekxk) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md]: Yes, I mean specific user which requested this feature. This is just an option first without `backward_weights`. I changed this description a little bit.

Comment (kwiersch) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md]: Okay, I finally got it. Thanks!

File: rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+| dnnl_use_scaleshift                            | *Inputs*: src, gamma, beta <br><br> *Outputs*: dst                      | *Inputs*: src, gamma, beta <br><br> *Outputs*: dst, mu, sigma^2                      | *Inputs*: diffdst, src, mu, sigma^2, gamma, beta <br><br> *Outputs*: diffsrc, diffgamma, diffbeta | Not supported                                                                                    |
+| dnnl_use_global_stats | dnnl_use_scaleshift    | *Inputs*: src, mu, sigma^2, gamma, beta <br><br> *Outputs*: dst         | *Inputs*: src, mu, sigma^2, gamma, beta <br><br> *Outputs*: dst                      | *Inputs*: diffdst, src, mu, sigma^2, gamma, beta <br><br> *Outputs*: diffsrc, diffgamma, diffbeta | Not supported                                                                                    |
```
[the remark this answers is not in the document]
Comment (bartekxk) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md]: Done thanks, my mistake with rebase.

File: rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md (hunk 1 above)
Comment (bartekxk) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md:8]: Still waiting for this information.

Comment (bartekxk) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md:8]: Ok, I received information, that they saw the crash while running a customer model but they do not recall now which model it was. They are still in favor of this change.

Comment (dzarukin) [reply rfcs/20210211-bnorm-backward_data-and-backward_weights/README.md:8]: Doesn't sound right to me as a primary motivator. I'd ask them to reproduce the issue and bring us VERBOSE output so that we could take a look since *there's no need in API extension to get what they want*.
